### PR TITLE
solution

### PR DESCRIPTION
**Pull request by ivan-shpynda**
[DEMO LINK](https://ivan-shpynda.github.io/react_static-list-of-posts/)

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -6,24 +6,19 @@ import posts from './api/posts';
...
-    <p>
-      <span>comments: </span>
-      {comments.length}
-    </p>
+const preparedPosts = posts.map(post => (
+  {
+    ...post,
+    user: users.find(user => user.id === post.userId),
+    comments: comments.filter(comment => comment.postId === post.id),
+  }
```
Comment (Grygoriy-Shytikov) [inline src/App.js:16]: Думаю {} тут зайві

File: src/components/User.jsx [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
+import { PropTypes } from 'prop-types';
+
+export function User({ user }) {
```
Comment (Grygoriy-Shytikov) [inline src/components/User.jsx:4]: ``` export function User({ name, email, address: { city, street, zipcode} }) {```

File: src/components/User.jsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  return (
+    <div className="user">
+      {'User: '}
+      {user.name}
```
Comment (Grygoriy-Shytikov) [inline src/components/User.jsx:8]: ``` {name}```

File: src/components/User.jsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+      {' Email: '}
+      {user.email}
```
Comment (Grygoriy-Shytikov) [inline src/components/User.jsx:10]: ```{email}```

File: src/components/User.jsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      <div>
+        {' Address: '}
+        {user.address.city}
```
Comment (Grygoriy-Shytikov) [inline src/components/User.jsx:13]: ```{city}```

File: src/components/User.jsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        {', '}
+        {user.address.street}
```
Comment (Grygoriy-Shytikov) [inline src/components/User.jsx:15]: ```{city}```

File: src/components/User.jsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        .
+        {' Zip Code: '}
+        {user.address.zipcode}
```
Comment (Grygoriy-Shytikov) [inline src/components/User.jsx:18]: ```{zipcode}```

**Review by Grygoriy-Shytikov — APPROVED** [review]
Виглядає добре 👍

На мою думку (можу помилятись) :
- Можна використати деструктуризацію:
  Наприклад як - дав пару коментарыв до User, відповідно в Post.jsx 11 строку замінити на ```<User {...user} />``` ...

- Перевірку типів важко читати

File: src/App.js (hunk 1 above)
Comment (VinnieJ-2k20) [reply src/App.js:16]: `{}` are necessary here because you create new object.

File: src/App.js [hunk 8]
```diff
@@ -6,24 +6,19 @@ import posts from './api/posts';
...
+));
 
-    <p>
-      <span>Users: </span>
-      {users.length}
-    </p>
+const App = () => (
+  <div className="App">
+    <PostList preparedPosts={preparedPosts} />
   </div>
```
Comment (VinnieJ-2k20) [inline src/App.js:22]: Page should have one `h1` header.

File: src/components/Comment.jsx [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  );
+}
+
+Comment.propTypes = {
+  comment: PropTypes.shape({
+    name: PropTypes.string.isRequired,
+    body: PropTypes.string.isRequired,
+    email: PropTypes.string.isRequired,
+  }).isRequired,
+};
```
Comment (VinnieJ-2k20) [inline src/components/Comment.jsx:32]: This is why we use spread operator when we pass object in props.

File: src/components/CommentList.jsx [hunk 10]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import { PropTypes } from 'prop-types';
+import { Comment } from './Comment';
+
+export function CommentList({ comments }) {
+  return (
+    <div>
+      {comments.map(comment => (
+        <Comment comment={comment} key={comment.id} />
```
Comment (VinnieJ-2k20) [inline src/components/CommentList.jsx:9]: Better to use spread operator here:
```suggestion
        <Comment {...comment} key={comment.id} />
```

File: src/components/CommentList.jsx [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  );
+}
+
+CommentList.propTypes = {
+  comments: PropTypes.arrayOf(PropTypes.shape({
+    body: PropTypes.string.isRequired,
+    email: PropTypes.string.isRequired,
+    id: PropTypes.number.isRequired,
+    name: PropTypes.string.isRequired,
+  })).isRequired,
```
Comment (VinnieJ-2k20) [inline src/components/CommentList.jsx:21]: No real need to validate body, email and id here.

File: src/components/User.jsx (hunk 2 above)
Comment (VinnieJ-2k20) [reply src/components/User.jsx:4]: It's better to avoid more than one level of destructuring.

**Review by VinnieJ-2k20 — APPROVED** [review]
I'll approve, but you should learn easier ways to operate objects in props.

**Pull request closed without merging**